### ISSUE TITLE
Inline formset fields have 'required' attribute (Django 1.10 issue)

### ISSUE THREAD

**Issue by adamcharnock** [django-polymorphic/django-polymorphic#251]
When using ``polymorphic_inlineformset_factory()`` it seems that the resulting forms have the ``required`` HTML attribute set, even though they should actually be optional.

I tracked this as far as the ``Form.use_required_attribute``, [which was added in Django 1.10](https://github.com/django/django/commit/ec6121693f112ae33b653b4364e812722d2eb567). This is now set in Django's own formset code, but isn't set by django-polymorphic.

However, I couldn't quite figure out how to actually fix this. I also only have so much time available so unfortunately I need to move on. I hope that this will help someone find a solution.

**Comment by ShaheedHaque** [comment]
I believe I just ran into this when attempting to upgrade from 2.0.5 to 2.1. Attempting to used the admin "add" on a polymorphic results in the following test failure:

```
self = <CompanyBankDetailGBForm bound=False, valid=Unknown, fields=(company;account_name;sort_code;account_number;reference;id1;id2;BACS_SUN)>
data = None, files = None, auto_id = 'id_%s'
prefix = 'companybankdetail_set-__prefix__', initial = {}
error_class = <class 'django.forms.utils.ErrorList'>, label_suffix = None
empty_permitted = True, field_order = None, use_required_attribute = None
renderer = None

    def __init__(self, data=None, files=None, auto_id='id_%s', prefix=None,
                 initial=None, error_class=ErrorList, label_suffix=None,
                 empty_permitted=False, field_order=None, use_required_attribute=None, renderer=None):
        self.is_bound = data is not None or files is not None
        self.data = {} if data is None else data
        self.files = {} if files is None else files
        self.auto_id = auto_id
        if prefix is not None:
            self.prefix = prefix
        self.initial = initial or {}
        self.error_class = error_class
        # Translators: This is the default suffix added to form field labels
        self.label_suffix = label_suffix if label_suffix is not None else _(':')
        self.empty_permitted = empty_permitted
        self._errors = None  # Stores the errors after clean() has been called.
    
        # The base_fields class attribute is the *class-wide* definition of
        # fields. Because a particular *instance* of the class might want to
        # alter self.fields, we create self.fields here by copying base_fields.
        # Instances should always modify self.fields; they should not modify
        # self.base_fields.
        self.fields = copy.deepcopy(self.base_fields)
        self._bound_fields_cache = {}
        self.order_fields(self.field_order if field_order is None else field_order)
    
        if use_required_attribute is not None:
            self.use_required_attribute = use_required_attribute
    
        if self.empty_permitted and self.use_required_attribute:
            raise ValueError(
>               'The empty_permitted and use_required_attribute arguments may '
                'not both be True.'
            )
E           ValueError: The empty_permitted and use_required_attribute arguments may not both be True.

../../../.local/lib/python3.6/site-packages/django/forms/forms.py:104: ValueError
```

Here, CompanyBankDetailGBForm is for a polymorphic base model CompanyBankDetail from which CompanyBankDetailGB is derived. I notice the link provided by @adamcharnock includes code which explicitly sets use_required_attribute in a couple of places. Based on that, I made the following patch:

```
$ diff -U5 /usr/local/lib/python3.6/dist-packages/polymorphic/formsets/models.py.saved /usr/local/lib/python3.6/dist-packages/polymorphic/formsets/models.py
--- /usr/local/lib/python3.6/dist-packages/polymorphic/formsets/models.py.saved 2018-05-14 08:34:10.680501699 +0100
+++ /usr/local/lib/python3.6/dist-packages/polymorphic/formsets/models.py       2018-08-28 10:33:33.979029587 +0100
@@ -138,10 +138,14 @@
         # BaseFormSet logic, with custom formset_class
         defaults = {
             'auto_id': self.auto_id,
             'prefix': self.add_prefix(i),
             'error_class': self.error_class,
+            # Don't render the HTML 'required' attribute as it may cause
+            # incorrect validation for extra, optional, and deleted
+            # forms in the formset.
+            'use_required_attribute': False,
         }
         if self.is_bound:
             defaults['data'] = self.data
             defaults['files'] = self.files
         if self.initial and 'initial' not in kwargs:
@@ -246,10 +250,11 @@
 
             form = form_class(
                 auto_id=self.auto_id,
                 prefix=self.add_prefix('__prefix__'),
                 empty_permitted=True,
+                use_required_attribute=False,
                 **kwargs
             )
             self.add_fields(form, None)
             forms.append(form)
         return forms
```

This seems to work for me, and I could prepare a PR myself but it is not clear to me what the test impact, if any, would be. Therefore it would be good if somebody knowledgeable could take this forward (I'm happy to assist as needed in that case).

**Comment by ShaheedHaque** [comment]
Never mind, I just noticed that #360 was fixed in 2.0.3. This issue can probably be closed.

**Comment by vdboor** [comment]
Thanks for mentioning it, yes, this issue was fixed in 2.0.3 :-)
